src/v/pandaproxy/rest/proxy.cc

```diff
@@ -22,51 +29,88 @@ namespace pandaproxy::rest {
 
 using server = ctx_server<proxy>;
 
+template<typename Handler>
+auto wrap(Handler h) {
+    return [h{std::move(h)}](
+             server::request_t rq,
+             server::reply_t rp) -> ss::future<server::reply_t> {
+        rq.authn_method = rq.service().config().authn_method(
+          rq.req->get_listener_idx());
+
+        if (rq.authn_method == config::rest_authn_method::http_basic) {
+            // Will throw if auth fails
+            // The catch statements in reply.h will do error handling
+            auto auth_result = rq.service().authenticator().authenticate(
+              *rq.req);
+            auth_result.pass();
+            rq.user = credential_t{
+              auth_result.get_username(), auth_result.get_password()};
+        }
+
+        co_return co_await h(std::move(rq), std::move(rp));
+    };
+}
+
 server::routes_t get_proxy_routes() {
     server::routes_t routes;
     routes.api = ss::httpd::rest_json::name;
 
     routes.routes.emplace_back(
-      server::route_t{ss::httpd::rest_json::get_brokers, get_brokers});
+      server::route_t{ss::httpd::rest_json::get_brokers, wrap(get_brokers)});
```

Comment:
@BenPope does this look safe? it looks like route_t is stashing a lambda coroutine and after the first time it is called its capture list would be destroyed?

Reply:
I think it's safe. It might be simpler to change the last line
```diff
- co_return co_await h(std::move(rq), std::move(rp));
+ return h(std::move(rq), std::move(rp));
```